google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,352 @@
+      std::shared_ptr<BackoffPolicy const> backoff_policy,
+      AsyncSleeper const sleeper,
+      AsyncStreamFactory<RequestType, ResponseType> stream_factory,
+      StreamInitializer<RequestType, ResponseType> initializer)
+      : retry_policy_prototype_(std::move(retry_policy)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        initializer_(std::move(initializer)) {}
+
+  ResumableAsyncStreamingReadWriteRpcImpl(ResumableStreamingReadRpc&&) = delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableStreamingReadRpc&&) = delete;
+
+  future<Status> Start() override {
+    future<Status> status_future;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      status_future = status_promise_.get_future();
+    }
+    Initialize(retry_policy_prototype_->clone(),
+               backoff_policy_prototype_->clone());
+    return status_future;
+  }
+
+  future<absl::optional<ResponseType>> Read() override {
+    future<absl::optional<ResponseType>> read_future;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (in_retry_loop_) {
+        read_promise_ =
+            absl::make_optional(promise<absl::optional<ResponseType>>());
+        return read_promise_.value().get_future();
+      }
+      read_future = stream_->Read();
+    }
+
+    return read_future.then(
+        [this](future<absl::optional<Response>> optional_response_future) {
+          auto optional_response = optional_response_future.get();
+          if (optional_response.has_value()) {
+            return make_ready_future(optional_response);
+          }
+          std::lock_guard<std::mutex> g{this->mu_};
+          read_promise_ =
+              absl::make_optional(promise<absl::optional<ResponseType>>());
+          if (!this->in_retry_loop_) {
+            this->OnFailure(kReadErrorStatus);
+            this->in_retry_loop_ = true;
+          }
+          return read_promise_.value().get_future();
+        });
+  }
+
+  future<bool> Write(Request const& r, grpc::WriteOptions o) override {
+    future<bool> write_future;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (in_retry_loop_) {
+        write_promise_ = absl::make_optional(promise<bool>());
+        return write_promise_.value().get_future();
+      }
+      write_future = stream_->Write(r, o);
+    }
+
+    return write_future.then([this](future<bool> write_fu) {
+      if (write_fu.get()) {
+        return make_ready_future(true);
+      }
+      std::lock_guard<std::mutex> g{this->mu_};
+      write_promise_ = absl::make_optional(promise<bool>());
+      if (!this->in_retry_loop_) {
+        this->OnFailure(kWriteErrorStatus);
+        this->in_retry_loop_ = true;
+      }
+      return write_promise_.value().get_future();
+    });
+  }
+
+  future<Status> Finish() override {
+    return stream_->Finish().then([this](future<Status> finish_future) {
```

Comment:
If stream_ failed with a permanent error, the retryer should have already called finish() on it